templates/environment/partials/vpc-endpoints.yml

```diff
@@ -0,0 +1,140 @@
+# Security group to allow receiving HTTPS traffic.
+VPCEndpointSecurityGroup:
+  Condition: CreateVPCEndpoints
+  Type: AWS::EC2::SecurityGroup
+  Properties:
+    GroupDescription: Security Group to allow use of VPC Endpoints.
+    VpcId: !Ref VPC
+    SecurityGroupIngress:
+      - IpProtocol: tcp
+        Description: HTTPS
+        FromPort: 443
+        ToPort: 443
+        CidrIp: 10.0.0.0/0
+
+# Allow ingress traffic from VPC endpoints to our containers.
+IngressFromVPCEndpoints:
+  Condition: CreateVPCEndpoints
+  Type: AWS::EC2::SecurityGroupIngress
+  Properties:
+    Description: Ingress from VPC endpoints.
+    GroupId: !Ref 'EnvironmentSecurityGroup'
+    IpProtocol: -1
+    SourceSecurityGroupId: !Ref 'VPCEndpointSecurityGroup'
+
+# Allow traffic from VPC endpoints.
+IngressFromItself:
+  Condition: CreateVPCEndpoints
+  Type: AWS::EC2::SecurityGroupIngress
+  Properties:
+    Description: Ingress from VPC endpoints.
+    GroupId: !Ref 'EnvironmentSecurityGroup'
+    IpProtocol: -1
+    SourceSecurityGroupId: !Ref 'EnvironmentSecurityGroup'
+
+{{- range $ind, $cidr := .PrivateSubnetCIDRs}}
+S3PrivateRouteTable{{inc $ind}}:
+  Type: AWS::EC2::RouteTable
+  Condition: CreateVPCEndpoints
+  Properties:
+    VpcId: !Ref 'VPC'
+
+S3PrivateRouteTable{{inc $ind}}Association:
+  Type: AWS::EC2::SubnetRouteTableAssociation
+  Condition: CreateVPCEndpoints
+  Properties:
+    RouteTableId: !Ref S3PrivateRouteTable{{inc $ind}}
+    SubnetId: !Ref PrivateSubnet{{inc $ind}}
+{{- end}}
+
+# ECS Resources
+ECSCluster:
```

Comment:
These were included in the prototype because the prototype was a single big cloudformation stack :) for Copilot some of these resources already exists